macros/core/PGML.pl

```diff
@@ -1266,19 +1266,22 @@ sub format {
 }
 
 sub string {
-	my $self    = shift;
-	my $block   = shift;
+	my ($self, $block) = @_;
+	my $stack   = $block->{stack};
+	my $n       = scalar(@$stack);
 	my @strings = ();
 	my $string;
-	foreach my $item (@{ $block->{stack} }) {
+	for my $i (0 .. $n - 1) {
+		my $item     = $stack->[$i];
+		my $next_par = $i + 1 < $n && $stack->[ $i + 1 ]{type} eq 'par' ? $stack->[ $i + 1 ] : undef;
```

Comment:
Change this to
```suggestion
	for my $i (0 .. $#$stack) {
		my $item     = $stack->[$i];
		my $next_par = $i < $#$stack && $stack->[ $i + 1 ]{type} eq 'par' ? $stack->[ $i + 1 ] : undef;
```
and delete `my $n = scalar(@$stack);` above (line 1271).  Perl is optimized for this and stores this number as a property of the array.  So there is no reason to save the array length to yet another location.  Also, perl offers the `$#` syntax for getting the index of the last entry in an array, so use that rather than adding or subtracting one from the array length.